File: resources/patterns/design/intercepting-filter-pattern/filter_chain.js
/**
 * Intercepting Filter pattern — runnable JavaScript example.
 *
 * A filter chain intercepts the request on the way in and the response on
 * the way back. Each filter can preprocess, short-circuit, or postprocess.
 *
 * Run: node filter_chain.js
 */

class HttpRequest {
  constructor(path, headers = {}) {
    this.path = path;
    this.headers = headers;
    this.user = null;
    this.authenticated = false;
  }
}

class HttpResponse {
  constructor() {
    this.status = 200;
    this.headers = {};
    this.body = null;
  }
}

class FilterChain {
  constructor() {
    this.filters = [];
    this.index = 0;
  }

  addFilter(filter) {
    this.filters.push(filter);
    return this;
  }

  doFilter(request, response) {
    if (this.index < this.filters.length) {
      const filter = this.filters[this.index++];
      filter.doFilter(request, response, this);
    }
  }

  // Entry point: reset the cursor so the same chain can serve the next request.
  execute(request, response) {
    this.index = 0;
    this.doFilter(request, response);
  }
}

class AuthenticationFilter {
  doFilter(request, response, chain) {
    const token = request.headers['Authorization'];
    if (token && token.startsWith('Bearer ')) {
      request.user = 'authenticated_user';
      request.authenticated = true;
      chain.doFilter(request, response);
    } else {
      response.status = 401;
      response.body = { error: 'Unauthorized' };
    }
  }
}

class LoggingFilter {
  doFilter(request, response, chain) {
    console.log(`[LOG] -> ${request.path}`);
    chain.doFilter(request, response);
    console.log(`[LOG] <- status ${response.status}`);
  }
}

class CompressionFilter {
  doFilter(request, response, chain) {
    chain.doFilter(request, response);
    const encoding = request.headers['Accept-Encoding'] || '';
    if (encoding.includes('gzip')) {
      response.headers['Content-Encoding'] = 'gzip';
      console.log('[COMPRESS] response compressed');
    }
  }
}

class TargetHandler {
  doFilter(request, response, chain) {
    if (response.status === 200) {
      response.body = { message: `Hello, ${request.user || 'guest'}!` };
    }
  }
}

function buildChain() {
  return new FilterChain()
    .addFilter(new AuthenticationFilter())
    .addFilter(new LoggingFilter())
    .addFilter(new CompressionFilter())
    .addFilter(new TargetHandler());
}

module.exports = {
  HttpRequest,
  HttpResponse,
  FilterChain,
  AuthenticationFilter,
  LoggingFilter,
  CompressionFilter,
  TargetHandler,
  buildChain,
};

if (require.main === module) {
  const chain = buildChain();

  const ok = new HttpRequest('/api/hello', {
    Authorization: 'Bearer abc123',
    'Accept-Encoding': 'gzip',
  });
  const denied = new HttpRequest('/api/hello', {});

  for (const req of [ok, denied]) {
    const res = new HttpResponse();
    chain.execute(req, res);
    console.log(`${req.path} auth=${req.authenticated} -> ${res.status} ${JSON.stringify(res.body)}\n`);
  }
}

File: resources/patterns/design/intercepting-filter-pattern/filter_chain.py
"""Intercepting Filter pattern — runnable Python example.

A filter chain intercepts the request on the way in and the response on
the way back. Each filter can preprocess, short-circuit, or postprocess.

Run:  python filter_chain.py
"""
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any


@dataclass
class HttpRequest:
    path: str
    headers: dict
    body: Any = None
    user: Any = None
    authenticated: bool = False


@dataclass
class HttpResponse:
    status: int = 200
    headers: dict = field(default_factory=dict)
    body: Any = None


class Filter(ABC):
    """Base filter that can delegate to the next filter."""

    def __init__(self, next_filter: "Filter" = None):
        self.next_filter = next_filter

    @abstractmethod
    def do_filter(self, request: HttpRequest, response: HttpResponse):
        pass

    def _invoke_next(self, request: HttpRequest, response: HttpResponse):
        if self.next_filter:
            self.next_filter.do_filter(request, response)


class AuthenticationFilter(Filter):
    """Short-circuits with 401 unless a Bearer token is present."""

    def do_filter(self, request: HttpRequest, response: HttpResponse):
        token = request.headers.get("Authorization")
        if token and token.startswith("Bearer "):
            request.user = "authenticated_user"
            request.authenticated = True
            self._invoke_next(request, response)
        else:
            response.status = 401
            response.body = {"error": "Unauthorized"}


class LoggingFilter(Filter):
    """Logs on both legs: request in, response out."""

    def do_filter(self, request: HttpRequest, response: HttpResponse):
        print(f"[LOG] -> {request.path}")
        self._invoke_next(request, response)
        print(f"[LOG] <- status {response.status}")


class CompressionFilter(Filter):
    """Postprocesses: marks the response as gzip-encoded if accepted."""

    def do_filter(self, request: HttpRequest, response: HttpResponse):
        self._invoke_next(request, response)
        if "gzip" in request.headers.get("Accept-Encoding", ""):
            response.headers["Content-Encoding"] = "gzip"
            print("[COMPRESS] response compressed")


class TargetHandler(Filter):
    """The final target: the actual business logic."""

    def __init__(self):
        super().__init__(None)

    def do_filter(self, request: HttpRequest, response: HttpResponse):
        if response.status == 200:
            response.body = {"message": f"Hello, {request.user or 'guest'}!"}


class FilterChain:
    """Builds the chain tail-to-head and executes it per request."""

    def __init__(self):
        self.filters: list[type[Filter]] = []

    def add_filter(self, filter_cls):
        self.filters.append(filter_cls)
        return self

    def execute(self, request: HttpRequest) -> HttpResponse:
        current = TargetHandler()
        for filter_cls in reversed(self.filters):
            f = filter_cls()
            f.next_filter = current
            current = f
        response = HttpResponse()
        current.do_filter(request, response)
        return response


if __name__ == "__main__":
    chain = FilterChain()
    chain.add_filter(AuthenticationFilter) \
         .add_filter(LoggingFilter) \
         .add_filter(CompressionFilter)

    ok = HttpRequest(
        path="/api/hello",
        headers={"Authorization": "Bearer abc123", "Accept-Encoding": "gzip"},
    )
    denied = HttpRequest(path="/api/hello", headers={})

    for req in (ok, denied):
        res = chain.execute(req)
        print(f"{req.path} auth={req.authenticated} -> {res.status} {res.body}\n")
